src/main/java/fi/aalto/cs/apluscourses/intellij/actions/TutorialAction.java

```diff
@@ -142,22 +142,25 @@ public boolean confirmCancel(@NotNull TutorialViewModel tutorialViewModel) {
     }
 
     @Override
-    public void end(@NotNull TutorialViewModel tutorialViewModel) {
-      JOptionPane.showMessageDialog(null,
+    public boolean finishAndSubmit(@NotNull TutorialViewModel tutorialViewModel) {
+      return JOptionPane.showConfirmDialog(null,
           getText("ui.tutorial.TutorialAction.end"),
           tutorialViewModel.getTitle(),
-          JOptionPane.INFORMATION_MESSAGE);
+          JOptionPane.YES_NO_OPTION) == JOptionPane.YES_OPTION;
     }
   }
 
-  private void onTutorialComplete(@NotNull MainViewModel mainViewModel) {
+  private void onTutorialComplete(@NotNull MainViewModel mainViewModel,
+                                  @NotNull Project project) {
     TutorialViewModel viewModel = mainViewModel.tutorialViewModel.get();
     if (viewModel != null) {
       viewModel.getTutorial().tutorialCompleted.removeCallback(mainViewModel);
+      if (viewModel.isCompleted() && dialogs.finishAndSubmit(viewModel)) {
+        new SubmitExerciseAction().submitTutorial(project, viewModel);
```

Comment:
This is what I meant in my previous comment. Actions are not designed to be created this way.